src/connectivity/components.c

```diff
@@ -425,32 +425,42 @@ int igraph_is_connected_weak(const igraph_t *graph, igraph_bool_t *res) {
     already_added[0] = 1;
     IGRAPH_CHECK(igraph_dqueue_push(&q, 0));
 
-    j = 1;
+    added_count = 1;
     while ( !igraph_dqueue_empty(&q)) {
-        long int actnode = (long int) igraph_dqueue_pop(&q);
         IGRAPH_ALLOW_INTERRUPTION();
-        IGRAPH_CHECK(igraph_neighbors(graph, &neis, (igraph_integer_t) actnode,
-                                      IGRAPH_ALL));
-        for (i = 0; i < igraph_vector_size(&neis); i++) {
+
+        long int actnode = (long int) igraph_dqueue_pop(&q);
+
+        IGRAPH_CHECK(igraph_neighbors(graph, &neis, (igraph_integer_t) actnode, IGRAPH_ALL));
+        long int nei_count = igraph_vector_size(&neis);
+
+        for (long int i = 0; i < nei_count; i++) {
             long int neighbor = (long int) VECTOR(neis)[i];
-            if (already_added[neighbor] != 0) {
+            if (already_added[neighbor]) {
                 continue;
             }
+
             IGRAPH_CHECK(igraph_dqueue_push(&q, neighbor));
-            j++;
-            already_added[neighbor]++;
+            added_count++;
+            already_added[neighbor] = 1;
+
+            if (added_count == no_of_nodes) {
+                /* We have already reached all nodes: the graph is connected. */
+                goto finish;
```

Comment:
I'd rather avoid the `goto` here and clear the dqueue explicitly instead and `break` out of the for loop. This would ensure that the `while` loop is also broken.